### ISSUE TITLE
go SDK/CGO support

### ISSUE THREAD

**Issue by hexfusion** [containers/bootc#522]
As the popularity of bootc grows a central location for go types/examples would be useful for projects. Speaking with @cgwalters he is willing to consider something similar to rpmostree-client-go[1] for bootc but there is some intersection with [2].

Curious if instead of creating go client as a os exec wrapper, exposing some of the Rust functionality via CGO? Please include me in future conversations on the topic if possible :pray: 

[1] https://github.com/coreos/rpmostree-client-go/
[2] https://github.com/containers/bootc/issues/518
[3] https://github.com/bytecodealliance/wasmtime/blob/b869b66bce8a57c9e83db4e3bcd6809d88e94a5f/crates/c-api/src/types/extern.rs#L39
[4]  https://github.com/bytecodealliance/wasmtime-go/blob/cabc733f95623fb897b2c35266124192ffca94b4/externtype.go#L9C12-L9C29

**Comment by cgwalters** [comment]
This issue also intersects with https://github.com/containers/bootc/issues/2

**Comment by cgwalters** [comment]
I've added this to the 1.2 milestone (next month) as it's pretty important for https://github.com/coreos/rpm-ostree/issues/4994

As of currently...I am slightly leaning towards varlink.

**Comment by cgwalters** [comment]
I looked a bit more at users of gRPC over Unix domain socket (what we definitely want here) and came upon https://github.com/hyperium/h2/pull/487 which is also pretty telling (in the opposite direction, most gRPC users are over TCP).

@sashagrunert any opinions on this?

**Comment by saschagrunert** [comment]
https://github.com/hyperium/h2/pull/487 is still unresolved and I personally think that's an issue. If using TCP is an alternative then gRPC seems more mature than Cap’n Proto. We still made good results using Cap’n Proto in conmon-rs, though. :)

**Comment by cgwalters** [comment]
We definitely don't want TCP here, it opens up huge questions around authorization, firewalling, lifecycle binding etc. even when operating just on loopback.

I like Cap'n Proto - they promote the speed thing but I think the fact that it's actually capability based (ref "security" on this page https://capnproto.org/rpc.html ) - I don't think that's true of gRPC and it's definitely not true of varlink. But the flip side is capnp is very complex (on the wire format and implementation).

I dunno...

**Comment by antheas** [comment]
In https://github.com/containers/bootc/pull/921 I introduce progress reporting via jsonl.

Essentially, JSONL is a protocol used in the NLP field to store large amounts of data in text files that can then be streamed.

The protocol consists of JSON objects, separated by new line characters.

In the PR, the caller provides a file descriptor that is used to stream little json packets that contain progress information. Then, those packets can be parsed by the caller using `jq` or the json parser of their choice to convert it into a progress bar. The format of the packets would be custom for bootc, although due to the tooling used for it it would be a low effort implementation for most callers.

The primary applications for this would be bash scripts and daemons that want to hook into bootc to provide basic update/installer functionality by calling it.

The non-ideal part of the PR is that due to the fact that bootc writes to both stdout and stderr indiscriminately, the 0 and 1 file descriptors cannot be used for this. Moreover, rust does not provide a way to change the stdout and stderr fds. This to me suggests that that a formatter lib should be used in lieu of `println` and `eprintln` to allow such indirection, but it is too late for that.

I would not consider this little format as the end all be all, but moreso as something to provide stopgap functionality until the bootc implementation matures. Even then, since such format introduces no dependencies and its implementation is tiny, it can stay there forever without introducing a maintenance burden.

**Comment by antheas** [comment]
I am not sure about embedding HTTP directly to bootc. Although it is something I did for [hhd](https://github.com/hhd-dev/hhd). 

For me, it was a strategic decision to allow binding react apps to it directly without a backend glue and to speedup, simplify the implementation of a Decky plugin and a local webapp that later became an appimage that is still used today. We got it running in two months. The main rival application to hhd used dbus, and still does not have a mature frontend application almost a year later. So it paid off for me.

HTTP has introduced some authentication woes for me that are not ideal. To allow frontends to consume it directly I also use a port for this, which requires HTTP auth, and I had to roll a semi-custom http server. As my implementation is more mature now, I have been thinking about offering an alternate IPC API, but I find myself asking the same question as this thread.

I would point to the fact that consuming it as a unix socket pretty much detracts 90% of its value, as you cant bind to it directly from a frontend and would need someone to rebroadcast it.

Something I'd suggest is regardless of the protocol used for this, the protocol itself should be "REST-like" so that it is easy to consume by a rootless process and rebroadcast to a TCP port, so that it can be consumed by javascript based frontends.

**Comment by antheas** [comment]
https://github.com/varlink/python/graphs/code-frequency does not inspire much confidence to me. gRPC looks more mature on the surface https://github.com/grpc/grpc but also more complex to use.

**Comment by cgwalters** [comment]
Right, bootc will never itself serve HTTP. Tools like [Cockpit](https://github.com/cockpit-project/cockpit/) exist and are highly mature ways to bridge from a local Unix to a remote browser frontend.
This issue is about how tools exactly like the cockpit bridge talk to bootc, not about any remote networking.

> https://github.com/varlink/python/graphs/code-frequency does not inspire much confidence to me. gRPC looks more mature on the surface https://github.com/grpc/grpc but also more complex to use.

Yeah...agreed.

**Comment by cgwalters** [comment]
FWIW I dug slightly more into the Rust varlink, and hit https://github.com/varlink/rust/issues/106 which isn't a showstopper...and just reading the code it is generally *good* code and overall well designed just yeah, not very actively maintained or contributed to.

As far as consuming it on the client side from some Python (or bash), what may actually work OK is using the `varlinkctl` client shipped by systemd and just parsing the output of *that* maybe?

Anyways...dunno. So short term I guess it's now looking like standardizing JSON output from our CLI and jsonl for progress, and if we feel pressure to do more then we can jump on either varlink or grpc.

**Comment by cgwalters** [comment]
It looks like varlink-rs is maintained (enough) and if I had to pick today I'd probably say let's go with that.

But again I don't have *extensive* experience (I do with dbus) and I have this tentative worry we could be painting ourselves into a corner.

The podman folks did a huge pivot away from varlink, it's worth noting. But OTOH, they had a hard constraint on implementing the Docker HTTP API as is which I suspect was the bigger thing.

**Comment by cgwalters** [comment]
In looking at protocols, I completely forgot about https://www.jsonrpc.org/ which is pretty widely used by e.g. https://microsoft.github.io/language-server-protocol/ and more recently https://modelcontextprotocol.io/introduction

varlink is perhaps like a json-rpc with some more opinions on IDL and tooling, but has the downside of being less widely used.

The only wrinkle still with both of these is that I see some use cases for fd passing which neither directly support. But the good thing here is that tweaking json-rpc to run over SOCK_SEQPACKET and define fd passing is pretty trivial and could be used to replace what's [used in skopeo](https://github.com/containers/skopeo/blob/main/docs-experimental/skopeo-experimental-image-proxy.1.md) today.
